### DISCUSSION TITLE
Agenda for ESMValTool community workshop: 19-21 November

### DISCUSSION THREAD

**Discussion by alistairsellar** [ESMValGroup/Community#192] [Meetings]
# Agenda ESMValTool Community Workshop 19-21 November 2024 (online meeting)
**All times in CET (UTC+1h)**

Teams links for participation will be emailed to registered participants. Complete this form by Friday 15  to register:  https://forms.office.com/e/2x34zAn2tD

### 19 November 2024 - Day 1 (Tuesday)
9.30 am – 9.45 am 	**Welcome, introductions, and workshop agenda** (Alistair)
9.45 am – 10.30 am       **New ESMValTool features** (Saskia)
10:30 am - 10:45 pm     **Documentation structure and ESMValTool/ESMValCore split** (TBD)

10:45 am – 11.00 am 	**Coffee break**

11.00 am - 11:30 am 	**Performance improvements. Discussion: community priorities for future improvements** (Bouwe)

11.30 am – 12:00 pm       **Governance and funding**
- Governance update - are we nearly there yet (Alistair)
- Discussion: Ongoing / upcoming projects and funding opportunities (Birgit) 

12.00 pm – 01.00 pm 	**Lunch break**
01.00 pm – 01.30 pm 	**Introduction of possible coding topics / sessions, e.g.** (Bouwe)
- Tutorial updates with User Engagement Team (Ranjini)
- Please add your proposals here...

01.30 pm – 03.30 pm 	**Coding session** (all)
03.30 pm – 03.45 pm	**Coffee break**
03.45 pm – 04.45 pm	**Coding session** (all)
04.45 pm – 05.00 pm	**Summary and wrap up of first day** (Birgit)

### 20 November 2024 - Day 2 (Wednesday)

10:00 am - 10:10 am   **Introduction day 2 and workshop agenda** (Birgit)

10:10 am – 10.50 am **Getting ready for CMIP7** (Alistair)
To include (before and after coffee break):
- Community requirements for ESMValTool to support analysis of CMIP7 (all)
- Resuming from a previous ESMValTool invocation (Kristi @k-a-webb)
- ESGF phase-out of CMIP3 and CMIP5 (Bouwe)
- CMIP Rapid Evaluation Framework (Birgit)

10:50 am - 11:00 am       **Filming for a promotional video** - (optional participation, Bouwe / NLeSC to introduce)
11.00 am – 11.15 am 	**Coffee break**

11.15 am – 12.00 am 	**CMIP7 continued** 

12.00 pm – 01.00 pm 	**Lunch break**

01.00 pm – 01.10 pm 	**Preparation for coding sessions** (Birgit)
01.10 pm – 03.00 pm 	**Coding session** (all)

03.00 pm – 03.15 pm 	**Coffee break**
03.15 pm – 05.00 pm 	**Coding session** (all)
05.00 pm – 05.15 pm 	**Summary and wrap up of second day** (Alistair)

**Virtual workshop dinner**: share a photo of your evening meal and/or beer!

### 21 November 2024 - Day 3 (Thursday)
9.30 am – 9.40 am 	**Introduction day 3 and workshop agenda** (Alistair)
9.40 am – 10.20 am  **CLIVAR ENSO metrics package & using ESMValCore and its preprocessing capabilities within a Jupyter notebook environment** (Felicity / Romain?)
10:20 am - 10:45 am **Updates to ESMValTool configuration format: design, progress and next steps** (Manuel)

10:45 am – 11.00 am 	**Coffee break**

11:00 am - 11:20 am **The future of NCL** (Manuel)

11:20 am – 12:00 pm **Technical infrastructure** (V)
- Update about using latest Python/numpy version
- Using `ruff` for formatting and linting ([intro](https://github.com/ESMValGroup/Community/discussions/192#discussioncomment-11206302))
- Do we like progress bars? (see e.g. [here](https://github.com/ESMValGroup/ESMValCore/pull/2522#issuecomment-2475712086))
- ...

12.00 pm – 01.00 pm 	**Lunch break**
01.00 pm – 01.10 pm 	**Preparation for coding sessions** (Alistair)
01.10 pm – 03.00 pm 	**Coding session** (all)
03.00 pm – 03.15 pm 	**Coffee break**
03.15 pm – 05.00 pm 	**Coding session** (all)
05.00 pm – 05.15 pm 	**Wrap up and future workshops** (Birgit)

**Comment by rswamina** [comment]
I would like to see a session to discuss some of the latest tool and core features. It would also be helpful to try out some of these if this is in a session so we can add the features to the tutorial. Some of these have been covered by @schlunma , @axel-lauer and @hb326 in talks given at various meetings including EGU this year. It would also be useful to discuss how to encourage the wider community to take advantage of these updates.

**Comment by rswamina** [comment]
I would like a coding session for Tutorial updates with the @ESMValGroup/userengagementteam as we have not met in a while.

**Comment by k-a-webb** [comment]
Related to discussion here, https://github.com/ESMValGroup/Community/discussions/187, I would appreciate a discussion about whether there is a plan to enable the use the pre-processed data from a previous ESMValTool run to as input to a new run. The are several limitations to the `resume_from` function (e.g., including additional datasets) which makes it unideal for developing monitoring or benchmarking recipes. I'd also like to hear from others about how they manage storing preprocessed data -- or is it operationally better to build then discard the NetCDF files for each run?

**Comment by schlunma** [comment]
Would be great to briefly talk about our new configuration (https://github.com/ESMValGroup/ESMValCore/issues/2371), in particular about already implemented features (https://github.com/ESMValGroup/ESMValCore/pull/2448) and potential next steps (e.g., https://github.com/ESMValGroup/ESMValCore/issues/2369).

**Comment by rbeucher** [comment]
We would like to explore using ESMValCore and its preprocessing capabilities within a Jupyter notebook environment. We are currently developing a series of Jupyter notebooks focused on ENSO evaluation, though they may not be fully ready for the workshop. However, we are more than happy to discuss the approach and gather feedback!

**Comment by bouweandela** [comment]
It would be nice to give a short presentation about the performance improvements we've made over the past year, e.g. https://github.com/ESMValGroup/ESMValCore/issues/674, https://github.com/ESMValGroup/ESMValCore/issues/2300, https://github.com/ESMValGroup/ESMValCore/issues/2301

**Comment by bouweandela** [comment]
We're creating some promotional videos about ESMValTool and would like to record a few scenes showing its lively community. Would it be possible to reserve 10 minutes for this on the agenda? Joining is completely optional of course.

**Comment by bouweandela** [comment]
There has been quite some interest in adding support for new data sources to ESMValCore, see https://github.com/ESMValGroup/ESMValCore/issues/2371, https://github.com/ESMValGroup/ESMValCore/issues/31, https://github.com/ESMValGroup/ESMValCore/pull/1894#issuecomment-1428667217, [@ESMValGroup/technical-lead-development-team meeting 2024-10-01](https://github.com/ESMValGroup/Community/blob/main/Tech%20Lead/Minutes/20241001.md). It would be nice to add a 1 hour breakout session somewhere in the agenda where interested people can join. People who showed interest so far are @bettina-gier @rbeucher @alistairsellar

**Comment by bouweandela** [comment]
@schlunma Would like to put the future of NCL on the agenda as discussed at the [@ESMValGroup/technical-lead-development-team meeting 2024-10-01](https://github.com/ESMValGroup/Community/blob/main/Tech%20Lead/Minutes/20241001.md)

**Comment by bouweandela** [comment]
The @ESMValGroup/technical-lead-development-team is considering switching to [ruff](https://docs.astral.sh/ruff/) for code formatting and linting. This has several advantages:
- ruff is the most popular tool at the moment, so it will be familiar for new users and lower the threshold to contributing
- we can use a single tool for everything, instead of many different tools (yapf, isort, prospector, flake8) see our [contribution docs](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#python) for more info on the current setup
- ruff is under much more active development than the collection of tools we currently use

 We're still waiting for `ruff` support in Codacy https://github.com/ESMValGroup/ESMValTool/issues/3739, but we would already like to announce this plan to the community and gather feedback.

**Reply by bouweandela** [reply]
We have already switched over ESMValCore to `ruff` formatting and enabled a few linting rules, but not yet fully replaced `prospector`.

**Comment by bouweandela** [comment]
Perhaps https://github.com/ESMValGroup/Community/discussions/187 would be a good topic for discussion?

**Comment by bouweandela** [comment]
And maybe https://github.com/ESMValGroup/Community/discussions/181?

**Comment by hb326** [comment]
I could talk a briefly about the CMIP Rapid Evaluation Framework (REF) since ESMValTool will be prominently represented there.

**Comment by valeriupredoi** [comment]
could I plss get a 30-odd minute slot to talk about technical infrastructure (devops-y) stuffs? Like: Python versions, deployment, ESMValBot, ruff and Codacy etc - a discussion of 10 min would be nice too, after. Danke much :beer:

**Comment by alistairsellar** [comment]
Thanks all for the proposals. Draft agenda now here: https://github.com/ESMValGroup/Community/discussions/192#discussion-7358231. Please comment.

@sloosvel @bouweandela @hb326 @rswamina @schlunma @k-a-webb @rbeucher @valeriupredoi - I have proposed you to either:
- chair a discussion
- present an update
- introduce a topic or question

Hopefully it is clear from the context and your suggestion which of these (chair / present / introduce) is required - shout if not. 

Please can you either confirm your name by removing the question mark, or reply to this comment if you don't think you're the right person. Feel free to swap roles with each other as you see fit.

**Reply by malininae** [reply]
Need to brew a giant coffee, but planning on attend the early sessions. I'll be the one with giant circles around the eyes :panda_face:

**Reply by k-a-webb** [reply]
I am happy to kick off the discussion of "Resuming from a previous ESMValTool invocation" as related to my work in model benchmarking, but I was not the original contributor of the issue I quoted (https://github.com/ESMValGroup/Community/discussions/187) and am not as familiar with the context of that request.

**Reply by alistairsellar** [reply]
Thanks @k-a-webb. I can support on the wider requirement for monitoring (my proposal in #187), but would be great to hear about your benchmarking needs and experience with `resume_from`.

**Comment by rbeucher** [comment]
I won't be able to make the 10:30AM on Tuesday. 
Sorry...

**Reply by alistairsellar** [reply]
Fair enough @rbeucher  - a bit late for you. There's no much scope to move it earlier, but 10:20 am Thursday is another option (though still late for you). I think you said that documentation split had been discussed in the @ESMValGroup/technical-lead-development-team - so is there anyone else who could introduce it?

**Reply by rbeucher** [reply]
I  already have plans for that night; otherwise, I would have joined. I'm not sure about anyone else, as we barely discussed it. Bouwe initiated the conversation, but I mentioned we could take on the work.

**Reply by alistairsellar** [reply]
Sorry about that. Perhaps one of us can present some ideas and gather feedback. Could you post a few bullets below on what you would propose and any particular questions for the feedback?

**Reply by rbeucher** [reply]
@bouweandela initiated this discussion. Here are my thoughts:

The current documentation is difficult to navigate and would benefit from a comprehensive review. I also see *ESMValCore* and *ESMValTool* as two distinct tools, each with unique roles and value:  

- **ESMValCore** is maintained by the core developers and adheres more closely to development standards. It is optimised, curated, and highly useful for reducing boilerplate code in model evaluation scripts. We’ve even begun using its processors in other Python workflows, demonstrating its broader applicability.  
- **ESMValTool**, on the other hand, evolves more organically through community contributions and serves a different purpose. A collection of diagnostics. 

I understand that the initial decision to brand both under a single name (*ESMValTool*, encompassing both ESMValCore and ESMValTool) was intended to minimise user confusion. However, I now believe that a clearer distinction between them would actually improve clarity. By highlighting their individual roles, we can better showcase the strengths of each tool while maintaining coherence in documentation and communication.

**Reply by alistairsellar** [reply]
Thanks @rbeucher - we used these points to kick off a discussion. The consensus was that separation would help clarity, though a note of caution that the signposting to ESMValCore for documentation on preprocessors needs to be really explicit, as not all new users will immediately understand the distinction.

All - as discussed in the session, please add your github handle here if you would be willing to contribute to the plan or implementation for these changes.

**Reply by bouweandela** [reply]
I would be happy to help with planning (and explaining the current setup).

**Reply by malininae** [reply]
I can also help in separating the docs and make the distinction clearer.

**Reply by ehogan** [reply]
I have been doing a lot of learning about documentation recently (for one of my other projects), so I would be keen to contribute to any planning / organising sessions 😊

**Reply by ehogan** [reply]
Shall we meet to plan splitting the ESMValTool documentation? Here is a poll to see if there's a good time for us to meet this month: https://www.when2meet.com/?27904063-ytAET. The dates run from next Tuesday, December 10 until Thursday, Dec 19 (when I go on leave until next year!). I realise it's short notice, but responses by early next week would be most useful! Otherwise we can try for January 😊

**Reply by ehogan** [reply]
@rbeucher, @flicj191, @malininae, would any of the dates in the above poll work for you? Or shall I create another poll for January? 😊

**Reply by rbeucher** [reply]
Hi @ehogan . I put my availability earlier today. Unfortunately it can only be evening for me next week...but go ahead with @flicj191 of that's easier.

**Reply by malininae** [reply]
Hi @ehogan thanks, I'm sorry, I was so behind on everything that I forgot to fill it! I indicated my times, if there will be a time that I'm not available, but works for everyone else, feel free to schedule the meeting without me.

**Reply by flicj191** [reply]
Hi @ehogan, my availability is in the poll as well, also the evenings of next week mainly.

**Reply by ehogan** [reply]
Ahh, silly me forgot to reload the page! 🙃 Thanks everyone! 👍 I'm going to suggest Thu, Dec 19 from 08:30 to 09:30 (apologies, edited to add: GMT), so that both @bouweandela and @malininae can attend some of the meeting. I hope that's ok? 😊

**Reply by ehogan** [reply]
Is a Teams meeting ok? If not, would someone setup a Zoom meeting, please? 😊

**Reply by rbeucher** [reply]
All good for me. @flicj191 ?

**Reply by malininae** [reply]
All works for me, thanks so much!

**Reply by flicj191** [reply]
yep sure thanks!

**Reply by bouweandela** [reply]
Thanks @ehogan! Could you please send a calendar invite to make sure we don't forget the meeting?

**Reply by ehogan** [reply]
Done! 🎉 If anyone else wants the invite please let me know! 😊

**Comment by alistairsellar** [comment]
Those who have or will present slides, please upload to https://github.com/ESMValGroup/Community/tree/main/Workshops/2024-11

(pinging @sloosvel @bouweandela @hb326 @rswamina @schlunma @k-a-webb @flicj191 @rbeucher @valeriupredoi as people named in the agenda, though you may not have slides)

**Comment by alistairsellar** [comment]
Wednesday's side meeting has been moved to Friday 9am CET (8 UTC). Please reply to this comment if you would like to join and don't yet have the meeting invite.

**Reply by sloosvel** [reply]
I'd be interested in joining

**Comment by malininae** [comment]
https://github.com/ESMValGroup/ESMValCore/issues/2582
